src/main/java/org/dmfs/jems/single/elementary/Digested.java

```diff
@@ -0,0 +1,106 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.single.elementary;
+
+import org.dmfs.iterables.decorators.Mapped;
+import org.dmfs.iterables.elementary.Seq;
+import org.dmfs.iterators.Function;
+import org.dmfs.jems.function.BiFunction;
+import org.dmfs.jems.function.elementary.SingleFunction;
+import org.dmfs.jems.messagedigest.MessageDigestFactory;
+import org.dmfs.jems.single.Single;
+
+import java.io.UnsupportedEncodingException;
+import java.security.MessageDigest;
+import java.util.Locale;
+
+
+/**
+ * A {@link Single} of a byte array which represents the digested value of the given input data.
+ *
+ * @author Marten Gajda
+ */
+public final class Digested implements Single<byte[]>
+{
+    private final MessageDigestFactory mMessageDigestFactory;
+    private final Iterable<Single<byte[]>> mParts;
+
+
+    public Digested(MessageDigestFactory messageDigestFactory, byte[]... parts)
+    {
+        this(messageDigestFactory, new Mapped<>(new Seq<>(parts), new SingleFunction<byte[]>()));
+    }
+
+
+    public Digested(MessageDigestFactory messageDigestFactory, CharSequence... parts)
+    {
+        this(messageDigestFactory, "UTF-8", parts);
+    }
+
+
+    public Digested(MessageDigestFactory messageDigestFactory, final String encoding, CharSequence... parts)
+    {
+        this(messageDigestFactory, new Mapped<>(new Seq<>(parts), new Function<CharSequence, Single<byte[]>>()
+        {
+            @Override
+            public Single<byte[]> apply(CharSequence bytes)
+            {
+                try
+                {
+                    return new ValueSingle<>(bytes.toString().getBytes(encoding));
+                }
+                catch (UnsupportedEncodingException e)
+                {
+                    throw new RuntimeException(String.format(Locale.ENGLISH, "%s encoding not supported by runtime", encoding), e);
+                }
+            }
+        }));
+    }
+
+
+    @SafeVarargs
+    public Digested(MessageDigestFactory messageDigestFactory, Single<byte[]>... parts)
+    {
+        this(messageDigestFactory, new Seq<>(parts));
+    }
+
+
+    public Digested(MessageDigestFactory messageDigestFactory, Iterable<Single<byte[]>> parts)
+    {
+        mMessageDigestFactory = messageDigestFactory;
+        mParts = parts;
+    }
+
+
+    @Override
+    public byte[] value()
+    {
+        return new Reduced<>(mMessageDigestFactory.newInstance(), new DigestFunction(), mParts).value().digest();
```

Comment:
I know this is the `Iterable` composing declarative way and that `Reduced` can possibly be used elsewhere as well, but this would be much easier to read for me:
```
        MessageDigest messageDigest = mMessageDigestFactory.newInstance();
        for (Single<byte[]> part : mParts)
        {
            messageDigest.update(part.value());
        }
        return messageDigest.digest();
```

Reply:
That's what I had initially. The pattern is quite common (we have it in http-client-essentials too) and I think it makes sense to write it like this. I think you'll get used to it ;-)